### PR TITLE
Async execution controls

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Closes https://github.com/enso-org/cloud-v2/issues/619
  - Async execution controls

### Important Notes
There is no design for this, as such, implementation details use placeholder designs.

- The context menu uses a play icon. An icon similar in style to the "copy" icon *may* work to represent "run in background", but it may be difficult to visually represent that it is being run in the background, without obstructing it with a foreground window
- The icon for projects being run in the background have a green tint, to distinguish them from projects that will be (or are currently) opened in the editor.
  - this will ***almost certainly*** need to be replaced with a proper design
- This *may* also make sense for the local backend, *however* as I don't know whether there is a way to access the completion progress of execution from the PM API, local backend support is currently *not* implemented in this PR.
  - On a related note: as far as I am aware, there is also no such endpoint for the cloud backend. However, support for async execution was recently added, so I am adding the basic functionality corresponding to the `executeAsync` project state.
- Whether a project is being run in the background is currently lost on refresh. This is because the async execution state is currently not sent by the backend.
- Placeholder shortcuts have been added (Shift+Enter - Cmd+Enter is already taken by the "share" action, and shift+double click). These are totally optional, and can easily be removed.

### Screenshots

New context menu entry:

image

Project being opened for async execution:

image

Project currently open for async execution:

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

**Comment by indiv0** [comment]
QA failed. One minor issue w/ stateful representation of projects found :)

# Scenario 0

```sh
./run gui watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
```

- ✅ GUI at http://localhost:8080 opens
- ✅ Can login
- ✅ Dashboard shows up
- ✅ Nodes in (non-Async) cloud project are visible

image

- ✅ New context menu entry "Run" is present

image

- ✅ Shift+Enter shortcut opens project in async mode
- ✅ Async run icon is visually distinct (i.e., green)

image

- ✅ Project "stop" icon is visually distinct for async projects

image

- ✅ Async projects can be closed like regular projects
- ❌ Async-running a project immediately after opening another one seems to cancel the in-progress open request

https://github.com/enso-org/enso/assets/641547/afa2496e-89bb-47ef-9735-4058f9c41a19

<blockquote><div><strong><a href="http://localhost:8080/">unfurl | built with Probot</a></strong></div></blockquote>

**Comment by somebody1234** [comment]
@indiv0 fix has been pushed

**Comment by PabloBuchu** [comment]
I think we shouldn't be able to open a scheduled project.. so please hide `up` arrow

Screenshot 2023-08-31 at 11 37 33

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/projectIcon.tsx [hunk 1]
```diff
@@ -368,15 +396,22 @@ export default function ProjectIcon(props: ProjectIconProps) {
...
-                        <SvgMask style={ICON_STYLE} src={StopIcon} />
+                        <SvgMask
+                            src={StopIcon}
+                            className={`${ICON_CLASSES} ${
+                                isRunningInBackground ? 'text-green' : ''
+                            }`}
+                        />
                     </button>
-                    {!isOtherUserUsingProject && (
+                    {!isOtherUserUsingProject && !shouldRunInBackground && (
```
Comment (PabloBuchu) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/projectIcon.tsx]: this `shouldRunInBackground` is out of scope

```
../dashboard/src/authentication/src/dashboard/components/projectIcon.tsx(414,51): error TS2304: Cannot find name 'shouldRunInBackground'.
```

**Review by PabloBuchu — CHANGES_REQUESTED** [review]

**Comment by somebody1234** [comment]
oops, was not at home then (and not at home rn) so couldn't(/can't) test atm

**Comment by PabloBuchu** [comment]
@somebody1234 how do we track projects that are running asyncronosuly? I am makng some improvements on backend side.. in the future when user click `run` the selected project will have status `Cloning` and in background we will create a child project which will eventually be opened with `async: true`

as on the screenshot below `New_project_1_cloned` is `opened` and has `execute_async` but the buttons are not greened and up arrow is visible.. 

Screenshot 2023-09-06 at 16 28 40

**Comment by somebody1234** [comment]
@PabloBuchu whoops lol... i was assuming the backend didn't send async state, but of course it does. currently i believe the frontend doesn't check that, but it should be simple enough to add that functionality.

re: the cloning stuff - i think it might need some more details on the lifecycle (and how the frontend should represent the lifecycle), because if the frontend needs to create a new row then that functionality will need to be specialcased - it shouldn't be hard, just noting that it'd need further details before it can be implemented

**Comment by PabloBuchu** [comment]
yeah I realized that state is not preserved after e.g reload

**Comment by PabloBuchu** [comment]
regarding the improvements lets not complicate things for now but easiest would be to reload directory in which the project was placed.. lets just fix what is within the scope of this PR.

**Comment by somebody1234** [comment]
@PabloBuchu updated. i did a tiny bit of testing - it seems like `opened_by` is not set, so upon refresh the stop button is disabled. not sure what's the best course of action here - one is to always enable it (= letting anyone stop the project), the other is to add `opened_by` to `execute_async` projects on the server. maybe it was just an oversight, or because `execute_async` was added before `opened_by` was added.

**Comment by somebody1234** [comment]
re: reloading the directory - currently the frontend avoids reloading the directory unless the backend or category is changed. this is partly to minimize server load, (i think) mostly because we switched to optimistic ui which kinda makes reloading unnecessary anyway, since we already have to simulate the result of the endpoint before it even reaches the server.

so basically: reloading the directory, while it *may* be an option, might not work well. if you think it's something that can be implemented in a separate PR, i think that's not a bad idea

**Review by PabloBuchu — APPROVED** [review]

**Pull request merged**